lib/active_model/serializer/adapter/json_api.rb

```diff
@@ -72,46 +87,56 @@ def fragment_cache(cached_hash, non_cached_hash)
           ActiveModel::Serializer::Adapter::JsonApi::FragmentCache.new.fragment_cache(root, cached_hash, non_cached_hash)
         end
 
-        protected
-
-        attr_reader :fieldset
-
         private
 
-        def serializable_hash_for_collection(options)
-          hash = { data: [] }
-          included = []
-          serializer.each do |s|
-            result = self.class.new(s, instance_options.merge(fieldset: fieldset)).serializable_hash(options)
-            hash[:data] << result[:data]
-            next unless result[:included]
+        # Build an array of resource objects for all requested resources (primary and included).
+        #
+        # @api private
+        def resource_objects_for(serializer, include_tree = @include_tree, is_primary = true, hashes = {})
+          if serializer.respond_to?(:each)
+            serializer.each { |s| resource_objects_for(s, include_tree, is_primary, hashes) }
+            return hashes
+          end
+
+          return hashes unless serializer && serializer.object
 
-            included |= result[:included]
+          key = [resource_identifier_type_for(serializer), resource_identifier_id_for(serializer)]
```

Comment:
gotchya.

Well, in any case, can you write in inline comments in the code about what's going on with that?